File: packages/python/minions_prompts/prompt_scorer.py
"""
PromptScorer — execute test cases against prompt versions and produce scored results.
"""

from __future__ import annotations

from typing import Any

from minions import Minion, create_minion, generate_id, now, Relation

from .storage import PromptStorage
from .schemas import prompt_result_type
from .prompt_renderer import PromptRenderer
from .types import TestRunResult, ComparisonResult


class PromptScorer:
    """Runs test cases against prompt versions and records scored results.

    Args:
        storage: The storage backend to use.

    Example::

        scorer = PromptScorer(storage)
        result = scorer.run_test(
            prompt_id,
            test_id,
            scores={"relevance": 85, "coherence": 90},
            passed=True,
        )
    """

    def __init__(self, storage: PromptStorage) -> None:
        self._storage = storage
        self._renderer = PromptRenderer()

    def run_test(
        self,
        prompt_id: str,
        test_id: str,
        *,
        scores: dict[str, float],
        passed: bool,
        output: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> TestRunResult:
        """Run a single test case against a prompt and record the result.

        Args:
            prompt_id: The ID of the prompt-template or prompt-version.
            test_id: The ID of the prompt-test minion.
            scores: Map of dimension names to scores (0–100).
            passed: Whether the test passes.
            output: Actual LLM output (if captured).
            metadata: Extra metadata (model, temperature, tokens, etc.).

        Returns:
            A TestRunResult with the created prompt-result minion.
        """
        prompt = self._storage.get_minion(prompt_id)
        test = self._storage.get_minion(test_id)
        if not prompt:
            raise ValueError(f"Prompt not found: {prompt_id}")
        if not test:
            raise ValueError(f"Test not found: {test_id}")

        content = (prompt.fields or {}).get("content", "") or ""
        input_variables = (test.fields or {}).get("inputVariables", {}) or {}

        rendered_prompt = self._renderer.render(str(content), dict(input_variables), strict=False)

        result_minion, _ = create_minion(
            {
                "title": f"Result: {test.title} on {prompt.title}",
                "fields": {
                    "renderedPrompt": rendered_prompt,
                    "output": output,
                    "scores": scores,
                    "metadata": metadata,
                    "passed": passed,
                },
            },
            prompt_result_type,
        )
        self._storage.save_minion(result_minion)

        # Create references
        for target_id in [test_id, prompt_id]:
            rel = Relation(
                id=generate_id(),
                source_id=result_minion.id,
                target_id=target_id,
                type="references",
                created_at=now(),
            )
            self._storage.save_relation(rel)

        return TestRunResult(
            prompt_id=prompt_id,
            test_id=test_id,
            rendered_prompt=rendered_prompt,
            scores=scores,
            passed=passed,
            result=result_minion,
        )

    def run_test_suite(
        self,
        prompt_id: str,
        test_ids: list[str],
        evaluations: list[dict[str, Any]],
    ) -> list[TestRunResult]:
        """Run multiple test cases against a prompt.

        Args:
            prompt_id: The ID of the prompt to test.
            test_ids: List of test-case IDs.
            evaluations: List of evaluation dicts with ``scores`` and ``passed`` keys.

        Returns:
            List of TestRunResult objects.
        """
        results = []
        for test_id, evaluation in zip(test_ids, evaluations):
            results.append(
                self.run_test(
                    prompt_id,
                    test_id,
                    scores=evaluation.get("scores", {}),
                    passed=evaluation.get("passed", False),
                    output=evaluation.get("output"),
                    metadata=evaluation.get("metadata"),
                )
            )
        return results

    def compare_versions(
        self,
        v1_id: str,
        v2_id: str,
        test_ids: list[str],
        v1_evaluations: list[dict[str, Any]],
        v2_evaluations: list[dict[str, Any]],
    ) -> list[ComparisonResult]:
        """Run A/B comparison between two versions against the same tests.

        Args:
            v1_id: ID of the baseline version.
            v2_id: ID of the comparison version.
            test_ids: List of test IDs to run against both.
            v1_evaluations: Evaluations for v1.
            v2_evaluations: Evaluations for v2.

        Returns:
            List of ComparisonResult objects.
        """
        v1_results = self.run_test_suite(v1_id, test_ids, v1_evaluations)
        v2_results = self.run_test_suite(v2_id, test_ids, v2_evaluations)

        comparisons = []
        for i, test_id in enumerate(test_ids):
            v1_result = v1_results[i]
            v2_result = v2_results[i]

            all_dims = set(v1_result.scores) | set(v2_result.scores)
            deltas = {
                dim: v2_result.scores.get(dim, 0) - v1_result.scores.get(dim, 0)
                for dim in all_dims
            }
            total_delta = sum(deltas.values())
            winner = "v2" if total_delta > 0 else "v1" if total_delta < 0 else "tie"

            comparisons.append(
                ComparisonResult(
                    v1_id=v1_id,
                    v2_id=v2_id,
                    test_id=test_id,
                    v1_result=v1_result,
                    v2_result=v2_result,
                    deltas=deltas,
                    winner=winner,
                )
            )
        return comparisons

File: packages/core/src/PromptScorer.ts
/**
 * @module minions-prompts/PromptScorer
 * Execute test cases against prompt versions and produce scored results.
 */

import { createMinion } from 'minions-sdk';
import type { Minion } from 'minions-sdk';
import type { PromptStorage, TestRunResult, ComparisonResult } from './types.js';
import { promptResultType } from './schemas.js';
import { PromptRenderer } from './PromptRenderer.js';

/**
 * Runs test cases against prompt versions and records scored results.
 *
 * @example
 * ```typescript
 * const scorer = new PromptScorer(storage);
 * const result = await scorer.runTest(promptId, testId, {
 *   scores: { relevance: 85, coherence: 90 },
 *   passed: true,
 * });
 * ```
 */
export class PromptScorer {
  private readonly renderer: PromptRenderer;

  constructor(private readonly storage: PromptStorage) {
    this.renderer = new PromptRenderer();
  }

  /**
   * Runs a single test case against a prompt and records the result.
   *
   * @param promptId - The ID of the prompt-template or prompt-version to test.
   * @param testId - The ID of the prompt-test minion to run.
   * @param evaluation - Scores and pass/fail provided by the caller.
   * @returns The test run result with the created prompt-result minion.
   */
  async runTest(
    promptId: string,
    testId: string,
    evaluation: {
      scores: Record<string, number>;
      passed: boolean;
      output?: string;
      metadata?: Record<string, unknown>;
    },
  ): Promise<TestRunResult> {
    const [prompt, test] = await Promise.all([
      this.storage.getMinion(promptId),
      this.storage.getMinion(testId),
    ]);

    if (!prompt) throw new Error(`Prompt not found: ${promptId}`);
    if (!test) throw new Error(`Test not found: ${testId}`);

    const promptFields = prompt.fields as Record<string, unknown>;
    const testFields = test.fields as Record<string, unknown>;
    const content = (promptFields['content'] as string) ?? '';
    const inputVariables = (testFields['inputVariables'] as Record<string, unknown>) ?? {};

    const renderedPrompt = this.renderer.render(content, inputVariables, { strict: false });

    const { minion: resultMinion } = createMinion(
      {
        title: `Result: ${test.title} on ${prompt.title}`,
        fields: {
          renderedPrompt,
          output: evaluation.output,
          scores: evaluation.scores,
          metadata: evaluation.metadata,
          passed: evaluation.passed,
        },
      },
      promptResultType,
    );

    await this.storage.saveMinion(resultMinion);

    // Create relations
    const { generateId, now } = await import('minions-sdk');
    await this.storage.saveRelation({
      id: generateId(),
      sourceId: resultMinion.id,
      targetId: testId,
      type: 'references',
      createdAt: now(),
    });
    await this.storage.saveRelation({
      id: generateId(),
      sourceId: resultMinion.id,
      targetId: promptId,
      type: 'references',
      createdAt: now(),
    });

    return {
      promptId,
      testId,
      renderedPrompt,
      scores: evaluation.scores,
      passed: evaluation.passed,
      result: resultMinion,
    };
  }

  /**
   * Runs multiple test cases against a prompt.
   *
   * @param promptId - The ID of the prompt to test.
   * @param testIds - Array of test-case IDs.
   * @param evaluations - Array of evaluations in the same order as testIds.
   * @returns Array of test run results.
   */
  async runTestSuite(
    promptId: string,
    testIds: string[],
    evaluations: Array<{
      scores: Record<string, number>;
      passed: boolean;
      output?: string;
      metadata?: Record<string, unknown>;
    }>,
  ): Promise<TestRunResult[]> {
    return Promise.all(
      testIds.map((testId, i) => this.runTest(promptId, testId, evaluations[i]!)),
    );
  }

  /**
   * Runs two prompt versions against the same tests and returns side-by-side comparison.
   *
   * @param v1Id - The ID of the first (baseline) version.
   * @param v2Id - The ID of the second (comparison) version.
   * @param testIds - Array of test IDs to run against both.
   * @param v1Evaluations - Evaluations for v1.
   * @param v2Evaluations - Evaluations for v2.
   * @returns Array of comparison results.
   */
  async compareVersions(
    v1Id: string,
    v2Id: string,
    testIds: string[],
    v1Evaluations: Array<{ scores: Record<string, number>; passed: boolean }>,
    v2Evaluations: Array<{ scores: Record<string, number>; passed: boolean }>,
  ): Promise<ComparisonResult[]> {
    const [v1Results, v2Results] = await Promise.all([
      this.runTestSuite(v1Id, testIds, v1Evaluations),
      this.runTestSuite(v2Id, testIds, v2Evaluations),
    ]);

    return testIds.map((testId, i) => {
      const v1Result = v1Results[i]!;
      const v2Result = v2Results[i]!;

      const allDimensions = new Set([
        ...Object.keys(v1Result.scores),
        ...Object.keys(v2Result.scores),
      ]);

      const deltas: Record<string, number> = {};
      let totalDelta = 0;

      for (const dim of allDimensions) {
        const d1 = v1Result.scores[dim] ?? 0;
        const d2 = v2Result.scores[dim] ?? 0;
        deltas[dim] = d2 - d1;
        totalDelta += deltas[dim]!;
      }

      const winner: ComparisonResult['winner'] =
        totalDelta > 0 ? 'v2' : totalDelta < 0 ? 'v1' : 'tie';

      return { v1Id, v2Id, testId, v1Result, v2Result, deltas, winner };
    });
  }
}
